framework/src/org/checkerframework/common/basetype/BaseTypeChecker.java

```diff
@@ -412,35 +425,181 @@ T getTypeFactoryOfSubchecker(Class<U> checkerClass) {
     // AbstractTypeProcessor delegation
     @Override
     public void typeProcess(TypeElement element, TreePath tree) {
+        if (getSubcheckers().size() > 0) {
+            messageStore = new TreeSet<>(checkerMessageComparator);
+        }
 
-        // If Java has issued errors, don't run any checkers on this compilation unit.
-        // If a sub checker issued errors, run the next checker on this compilation unit.
-
-        // Log.nerrors counts the number of Java and checker errors have been issued.
-        // super.typeProcess does not typeProcess if log.nerrors > errorsOnLastExit
+        // Errors (or other messages) issued via
+        // SourceChecker#message(Diagnostic.Kind, Object, String, Object...)
+        // are stored in messageStore until all checkers have processed this compilation unit.
+        // All other messages are printed immediately.  This includes errors issued because the
+        // checker threw an exception or called ErrorReporter.errorAbort().
 
         // In order to run the next checker on this compilation unit even if the previous
         // issued errors, the next checker's errsOnLastExit needs to include all errors
         // issued by previous checkers.
 
-        // To prevent any checkers from running if a Java error was issued for this compilation unit,
-        // errsOnLastExit should not include any Java errors.
         Context context = ((JavacProcessingEnvironment) processingEnv).getContext();
         Log log = Log.instance(context);
-        // Start with this.errsOnLastExit which will account for errors seen by
-        // by a previous checker run in an aggregate checker.
+
         int nerrorsOfAllPreviousCheckers = this.errsOnLastExit;
-        for (BaseTypeChecker checker : getSubcheckers()) {
-            checker.errsOnLastExit = nerrorsOfAllPreviousCheckers;
+        for (BaseTypeChecker subchecker : getSubcheckers()) {
+            subchecker.errsOnLastExit = nerrorsOfAllPreviousCheckers;
+            subchecker.messageStore = messageStore;
             int errorsBeforeTypeChecking = log.nerrors;
 
-            checker.typeProcess(element, tree);
+            subchecker.typeProcess(element, tree);
 
             int errorsAfterTypeChecking = log.nerrors;
             nerrorsOfAllPreviousCheckers += errorsAfterTypeChecking - errorsBeforeTypeChecking;
         }
+
         this.errsOnLastExit = nerrorsOfAllPreviousCheckers;
         super.typeProcess(element, tree);
+
+        if (getSubcheckers().size() > 0) {
+            printCollectedMessages(tree.getCompilationUnit());
+            this.errsOnLastExit = log.nerrors;
```

Comment:
Why is this change needed? Maybe some adapted version of the removed comments (lines 426/430) would help explain?